File: back/src/main/java/com/back/domain/ticket/listener/TicketCancelledEventHandler.java
package com.back.domain.ticket.listener;

import com.back.domain.ticket.event.TicketCancelledEvent;
import com.back.domain.waiting.service.WaitingQueueService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

@Component
@RequiredArgsConstructor
public class TicketCancelledEventHandler {

    private final WaitingQueueService waitingQueueService;

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    public void handleTicketCancelled(TicketCancelledEvent event) {
        waitingQueueService.allowEntry(event.concertId(), event.scheduleId());
    }
}

File: back/src/main/kotlin/com/back/domain/ticket/listener/TicketCancelledEventHandler.kt
package com.back.domain.ticket.listener

import com.back.domain.ticket.event.TicketCancelledEvent
import com.back.domain.waiting.service.WaitingQueueManager
import com.back.domain.waiting.service.WaitingQueueService
import org.springframework.stereotype.Component
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Component
class TicketCancelledEventHandler(
    private val waitingQueueService: WaitingQueueService,
    private val waitingQueueManager: WaitingQueueManager
) {

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT, fallbackExecution = true)
    fun handleTicketCancelled(event: TicketCancelledEvent) {
        waitingQueueManager.removeActiveUser(event.scheduleId, event.userId)
        waitingQueueService.allowEntry(event.concertId, event.scheduleId)
    }
}
